### PR TITLE
Add an ability to locate nodes using computed accessibility attributes

### PR DESCRIPTION
**Pull request by OrKoN**
Based on: https://github.com/w3c/webdriver-bidi/pull/624
Issue: https://github.com/w3c/webdriver-bidi/issues/443

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver-bidi/pull/660.html" title="Last updated on Apr 15, 2024, 12:44 PM UTC (b0ee307)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver-bidi/660/efb27e8...b0ee307.html" title="Last updated on Apr 15, 2024, 12:44 PM UTC (b0ee307)">Diff</a>

### CODE REVIEW

File: index.bs
Comment (OrKoN) [reply index.bs:2218]: Extensible is needed to allow implementation extension attributes

Comment (sadym-chromium) [reply index.bs:2218]: In sucha case we have to extend the *locate nodes using accessibility attributes* algorithm with mentioning that. Current specification version says that the filter should respect only `name` and `role`, and any other fields would be ignored. If the intention is to allow some vendor-specific filtering, I guess it should be mentioned here.

Comment (OrKoN) [inline index.bs:2218]:
```suggestion
```

Comment (OrKoN) [reply index.bs]: Removed extensible for now.

**Review by sadym-chromium — APPROVED** [review]

**Comment by OrKoN** [comment]
@jrandolf PTAL

File: index.bs
Comment (whimboo) [inline index.bs:2213]: nit: Please move up before `CssLocator` to keep the alphabetical order. Same for `browsingContext.Locator` as well.

File: index.bs
Comment (whimboo) [inline index.bs:3107]: Not sure I understand that logic here. Shouldn't we first run some kind of selector to find specific nodes below the start nodes? Note that `context nodes` are these start nodes. And right now we would just return these which have the wanted role and name.

So shouldn't it be similar to the CSS selector but by using the `[aria-label=%label%]`.

Comment (OrKoN) [reply index.bs:3107]: I believe iteration happens here:

```
  1. Let |child nodes| be an empty [=/list=] and, for each node |child| in the
     <a spec=dom>children</a> of |context node|:

    1. If |child| implements {{Element}}, [=list/append=] |child| to |child nodes|.
```

so this algorithm returns all nodes that match the predicate (role OR name matches). aria-label is not a computed attribute and, therefore, does not query in the same way.

File: index.bs
Comment (whimboo) [inline index.bs:2196]: Please move this line up as well.

File: index.bs
Comment (whimboo) [inline index.bs:2217]:
```suggestion
    ? name: text,
    ? role: text,
```

Beside that one more question... In WebDriver classic we have [`Get Computed Label`](https://w3c.github.io/webdriver/#get-computed-label), so should we use `label` instead of `name`?

File: index.bs
Comment (whimboo) [reply index.bs:3107]: You are right. Interesting that I missed that.

Beside that one question. For the `locate nodes using inner text` locator we also have a `maxDepth` optional argument to limit the search on the DOM. I wonder if we should have it here as well given that computing the role and label might be expensive as well, or is that not the case?

File: index.bs
Comment (whimboo) [inline index.bs:3140]: When seeing these lines again I actually wonder if that behavior is correct. Consider a node with 2 children whereby the first children has a lot of children as well. If we have collected more matching nodes as allowed, the 2nd child of the original context node wouldn't be included.

I wonder if we should flatten the array at the very end from top to bottom to not actually loose important already collected nodes. It's not blocking this PR from landing but we should at least file a new issue to further discuss (if necessary) and fix.

**Review by whimboo — CHANGES_REQUESTED** [review]

File: index.bs
Comment (OrKoN) [reply index.bs:3107]: I am not sure: for now it was not an issue in Puppeteer but we could add it later if it becomes a problem (usually those attributes are cached for a11y purposes anyway). There is `maximum returned node count` to limit the number of matches returned and abort early and that helps with the expensiveness of the search.

File: index.bs
Comment (OrKoN) [reply index.bs:3140]: I think you are right and it looks like the innerText selector has the same issue, let me look over it again

File: index.bs
Comment (OrKoN) [reply index.bs:2217]: I am not sure why it is called label since it references accessible name in the spec (the spec name is "Accessible Name and Description Computation 1.1")

File: index.bs
Comment (OrKoN) [reply index.bs:3444]: Rewrote this part.

File: index.bs
Comment (whimboo) [reply index.bs]: I can see: 

> The purpose of the computation is to create a [perceivable](https://www.w3.org/TR/accname-1.1/#dfn-perceivable) label or description for alternative presentations, in the form of a flat string of space separated textual tokens.

@jgraham what's your take on that?

File: index.bs
Comment (whimboo) [reply index.bs]: I reopened this issue for awaiting the results of the check.

Comment (OrKoN) [reply index.bs]: I have rewrote that part also addressing this issue (marked wrong thread).

**Review by sadym-chromium — APPROVED** [review]

File: index.bs
Comment (juliandescottes) [reply index.bs]: I would be in the favor of name as well, since this is about accessible name. The accessible name might be derived from an aria-label, but not necessarily (can come from title etc...).
Also, both Firefox and Chrome DevTools expose this as "name" in the UI, so it could be helpful for users writing tests if it matches our APIs.

Comment (whimboo) [reply index.bs]: Ah good to know. Aligning what DevTools has might actually be also a good idea and will help web developers and testers to align

File: index.bs
Comment (jgraham) [inline index.bs:3427]: It technically doesn't matter, but I don't love this organisation where we recursively call into the algorithm and re-perform steps that are actually redundant.

I think I'd prefer that we pull out the parts that actually make sense to run in each recursive call into their own algorithm and just call that.

File: index.bs
Comment (jgraham) [inline index.bs:3461]: So we can end up with more than `|maximum returned node count|` nodes being returned? That seems to violate the assert in the calling steps. I think we either need to check each time we append, or truncate the list at the end.

**Review by jgraham — CHANGES_REQUESTED** [review]

**Review by christian-bromann — APPROVED** [review]
This will be very helpful for WebdriverIO not have to create weird xpath selectors that aren't 100% correct all the time 

👍

**Comment by lutien** [comment]
@OrKoN, are you planning to add wpt tests for it, once the spec is merged?

**Comment by OrKoN** [comment]
@lutien yes

File: index.bs
Comment (whimboo) [inline index.bs:3434]: Should this by default really be true? It would mean that if a selector is used that doesn't contain any `name` or `role` we would match all the nodes in the document. This could be kinda invasive for huge documents. Or can a computed role / name never be `undefined`?

Nevertheless isn't it better to set it to `false` here by default and to `true` only when a match has found?

File: index.bs
Comment (whimboo) [inline index.bs:3440]: Not sure how complex the computation of both the role and name are, but we won't have to do it when the given key hasn't been specified, right?

**Review by jgraham — CHANGES_REQUESTED** [review]

File: index.bs
Comment (OrKoN) [reply index.bs:3434]: That is checked in the beginning of the main algorithm, so at least one attribute will be provided. Setting default to false makes the text more verbose as it cannot be flipped to true based on a single attribute check.

File: index.bs
Comment (OrKoN) [reply index.bs:3440]: yes, I believe the key is checked above at line 3436

File: index.bs
Comment (jgraham) [reply index.bs:3457]: I mean that it shouldn't say "is greater [than] or equal to" if it's an invariant that it can't be greater than, which the later assert requires that it is. So I think we should remove "if greater or".

Comment (OrKoN) [reply index.bs:3457]: Also, I think a break would continue the recursion so it has to be a return to prevent redundant iterations?

Comment (jgraham) [reply index.bs:3457]: Unless I'm misunderstanding the semantics there's only one loop here? And neither break nor return will propogate through recursive calls. Anyway either is fine.

File: index.bs
Comment (jgraham) [inline index.bs:3469]: This is indented too much, right? i.e. the return should be after the loop, otherwise we only ever process the first context node and the first element children below it.

**Review by jgraham — CHANGES_REQUESTED** [review]

File: index.bs
Comment (OrKoN) [reply index.bs]: Yeah, I have misread the indentation

File: index.bs
Comment (jgraham) [inline index.bs:3565]:
```suggestion
         1. Let |result nodes| be [=locate nodes using accessibility attributes=]
```

Since we're no longer returning a result type, we can't use `[=trying=]`.

**Review by jgraham — APPROVED** [review]
But note the small fixup required.

**Pull request merged**
